app/system/ldap/ldap.py

```diff
@@ -0,0 +1,77 @@
+import logging
+import ldap3
+
+from .config import LDAPConfig
+
+logger = logging.getLogger(__name__)
+
+
+class LDAP():
+    config: LDAPConfig
+    server: ldap3.Server
+
+    def __init__(self, config: LDAPConfig):
+        self.config = config
+        self.server = ldap3.Server(
+            config.ldap_server, config.ldap_port, use_ssl=config.ldap_use_ssl)
+
+    def set_password(self, username, password):
+        conn = ldap3.Connection(self.server, user=self.config.admin_user,
+                                password=self.config.admin_password, version=3)
+        if not conn.bind():
+            logger.error('LDAP bind failed with user %s',
+                         self.config.admin_user)
+            return None
+
+        _, _, distinguished_name = self.getuser(username)
+        return conn.extend.microsoft.modify_password(distinguished_name, password)
+
+    def getuser(self, user_principal_name, search_base=None):
+        if search_base is None:
+            search_base = self.config.user_base
+
+        conn = ldap3.Connection(self.server, user=self.config.admin_user,
+                                password=self.config.admin_password, read_only=True, version=3)
+        if not conn.bind():
+            logger.error('LDAP bind failed with user %s',
+                         self.config.admin_user)
+            return None
+
+        user_principal_name = self.normalize_login(user_principal_name)
+        search_filter = f'(&(objectCategory=person)(objectClass=user)(userPrincipalName={user_principal_name}))'
+        attrs = ['userPrincipalName', 'displayName', 'distinguishedName']
+        if not conn.search(search_base, search_filter, attributes=attrs):
+            logger.error(
+                'No LDAP entries found for search base %s and filter %s', search_base, search_filter)
+            return None
```

Comment:
In Python, returning None can be ambiguous and unclear as to why it was returned. It is usually better to raise an exception with a clear message that indicates why the function failed or to return a value that accurately reflects the result of the function.